File: src/ui_dialogs.cpp
// ui_dialogs.cpp

#include "ui_dialogs.h"
#include <cstdlib>
#include <cwchar>
#include <vector>

// Windows-Specific Code
#include <windows.h>
#include <windowsx.h>
#include <shobjidl.h>
// End Windows-Specific Code

using namespace std;

#ifndef MIN
#define MIN(a, b) ((a) < (b) ? (a) : (b))
#endif

// More Windows Stuff
bool init_com_sta() {
    HRESULT hr = CoInitializeEx(NULL, 
        COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
    return SUCCEEDED(hr);
}

// REF: 
// https://docs.microsoft.com/en-us/windows/win32/learnwin32/example--the-open-dialog-box
// https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb776913(v%3Dvs.85)
// https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/dd940349%28v%3dvs.85%29
// http://codexpert.ro/blog/2013/05/21/file-open-dialog-with-multiple-selection-part-3-cutting-the-dog-tail/
// https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb776913(v=vs.85)
void ShowUserOpenFileDialog(IFileOpenDialog *pFileOpen, char *nameBuffer, 
    int bufferSize, int *numChars)
{
    // Show the dialog
    HRESULT hr = pFileOpen->Show(NULL);

    // Get the filename from the dialog
    if (SUCCEEDED(hr)) {
        IShellItem *pItem;
        hr = pFileOpen->GetResult(&pItem);
        if (SUCCEEDED(hr)) {
            PWSTR pszFilePath;
            hr = pItem->GetDisplayName(SIGDN_FILESYSPATH, &pszFilePath);
            
            // Copy the filename to the output buffer
            size_t lstr = wcslen(pszFilePath);
            *numChars = (int)wcstombs(nameBuffer, pszFilePath, MIN(bufferSize, lstr));

            // Clean Up
            CoTaskMemFree(pszFilePath);
        }
        pItem->Release();
    }
}



void ShowUserMultiFileOpenDialog(IFileOpenDialog *pFileOpen, char **nameBuffers, 
    int numBuffers, int *numNames, int bufferSize, int *numChars)
{
    // Show the dialog
    HRESULT hr = pFileOpen->Show(NULL);

    // Get the filenames from the dialog
    if (SUCCEEDED(hr)) {
        // See https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb776913(v=vs.85)
        IShellItemArray *rsts;
        hr = pFileOpen->GetResults(&rsts);
        if (SUCCEEDED(hr)) {
            // Retrieve the results
            DWORD nItems;
            rsts->GetCount(&nItems);
            *numNames = MIN((int)nItems, numBuffers);
            for (int i = 0; i < *numNames; ++i) {
                IShellItem *pItem;
                hr = rsts->GetItemAt((DWORD)i, &pItem);
                if (SUCCEEDED(hr)) {
                    PWSTR pszFilePath;
                    hr = pItem->GetDisplayName(SIGDN_FILESYSPATH, &pszFilePath);

                    size_t lstr = wcslen(pszFilePath);
                    numChars[i] = (int)wcstombs(nameBuffers[i], pszFilePath, MIN(bufferSize, lstr));

                    CoTaskMemFree(pszFilePath);
                }
                pItem->Release();
            }

            // Clean Up
            rsts->Release();
        }
    }
}



vector<COMDLG_FILTERSPEC> ToFilterSpec(int nfilters, const char **filterName, 
    const char **filterPattern)
{
    const int BUFFER_SIZE = 2048;
    vector<COMDLG_FILTERSPEC> list(nfilters);
    wchar_t *nameBuffer, *patternBuffer;

    for (int i = 0; i < nfilters; ++i) {
        nameBuffer = new wchar_t[BUFFER_SIZE];
        size_t nchars = mbstowcs(nameBuffer, filterName[i], BUFFER_SIZE - 1);

        patternBuffer = new wchar_t[BUFFER_SIZE];
        nchars = mbstowcs(patternBuffer, filterPattern[i], BUFFER_SIZE - 1);

        COMDLG_FILTERSPEC spec = {nameBuffer, patternBuffer};
        list[i] = spec;
    }
    
    return list;
}

void CleanUpFilterSpecList(vector<COMDLG_FILTERSPEC> &list) {
    for (size_t i = 0; i < list.size(); ++i) {
        auto item = list[i];
        delete [] item.pszName;
        delete [] item.pszSpec;
    }
}



void ShowUserSaveDialog(IFileSaveDialog *dlg, char *name, int bufferSize, int *numChars) {
    // Show the dialog
    HRESULT hr = dlg->Show(NULL);

    // Get the filename
    if (SUCCEEDED(hr)) {
        IShellItem *pItem;
        hr = dlg->GetResult(&pItem);
        if (SUCCEEDED(hr)) {
            PWSTR filePath;
            hr = pItem->GetDisplayName(SIGDN_FILESYSPATH, &filePath);

            // Copy the filename to the output buffer
            size_t lstr = wcslen(filePath);
            *numChars = (int)wcstombs(name, filePath, MIN(bufferSize, lstr));

            // Clean Up
            CoTaskMemFree(filePath);
        }
        pItem->Release();
    }
}

// End of Windows Stuff


bool init_ui_env_c() 
{
    // Windows-Specific Call
    return init_com_sta();
    // End Windows-Specific Call
}



void clean_up_ui_env_c() 
{
    // Windows-Specific Call
    CoUninitialize();
    // End Windows-Specific Call
}



void show_open_file_dialog_c(char *nameBuffer, int bufferSize, int *numChars, 
    int nfilters, const char **filterName, const char **filterPattern)
{
    // Initialize outputs
    *numChars = 0;

    // Windows-Specific Code
    
    // Create the FileOpenDialog object
    IFileOpenDialog *pFileOpen;
    HRESULT hr = CoCreateInstance(CLSID_FileOpenDialog, NULL, 
        CLSCTX_ALL, IID_IFileOpenDialog, 
        reinterpret_cast<void**>(&pFileOpen));
    if (SUCCEEDED(hr)) {
        // Set the filter options
        if (nfilters > 0) {
            auto filters = ToFilterSpec(nfilters, filterName, filterPattern);
            hr = pFileOpen->SetFileTypes(filters.size(), filters.data());
            if (SUCCEEDED(hr)) {
                // Show the dialog
                ShowUserOpenFileDialog(pFileOpen, nameBuffer, bufferSize, 
                    numChars);
            }
            CleanUpFilterSpecList(filters);
        }
        else {
            // No filters - just show the dialog
            ShowUserOpenFileDialog(pFileOpen, nameBuffer, bufferSize, numChars);
        }
        pFileOpen->Release();
    }
    // End Window-Specific Code
}



void show_multi_file_open_file_dialog_c(char **nameBuffers, int numBuffers, 
    int bufferSize, int *numNames, int *numChars, int nfilters, 
    const char **filterName, const char **filterPattern)
{
    // Initialize outputs
    *numNames = 0;

    // Windows-Specific Code

    // Create the FileOpenDialog object
    IFileOpenDialog *pFileOpen;
    HRESULT hr = CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_ALL, 
        IID_IFileOpenDialog, reinterpret_cast<void**>(&pFileOpen));
    if (SUCCEEDED(hr)) {
        DWORD dwOptions;
        hr = pFileOpen->GetOptions(&dwOptions);
        if (SUCCEEDED(hr)) {
            hr = pFileOpen->SetOptions(dwOptions | FOS_ALLOWMULTISELECT);
        }

        // Set the filter options
        if (nfilters > 0) {
            auto filters = ToFilterSpec(nfilters, filterName, filterPattern);
            hr = pFileOpen->SetFileTypes(filters.size(), filters.data());
            if (SUCCEEDED(hr)) {
                ShowUserMultiFileOpenDialog(pFileOpen, nameBuffers, numBuffers, 
                    numNames, bufferSize, numChars);
            }
            CleanUpFilterSpecList(filters);
        }
        else {
            // No filters - just show the dialog
            ShowUserMultiFileOpenDialog(pFileOpen, nameBuffers, numBuffers, 
                numNames, bufferSize, numChars);
        }
        pFileOpen->Release();
    }
    // End Window-Specific Code
}



void show_save_file_dialog_c(const char *defaultExt, int nfilters, 
    const char **filterName, const char **filterPattern, char *name,
    int bufferSize, int *numChars)
{
    // Initialization
    *numChars = 0;

    // Windows-Specific Code

    // Create the dialog
    IFileSaveDialog *dlg;
    HRESULT hr = CoCreateInstance(CLSID_FileSaveDialog, NULL, CLSCTX_ALL, 
        IID_IFileSaveDialog, reinterpret_cast<void**>(&dlg));

    // Set the default file extension
    size_t n = strlen(defaultExt) + 1;
    wchar_t *ext = new wchar_t[n + 1];
    mbtowc(ext, defaultExt, n);
    dlg->SetDefaultExtension(ext);
    delete [] ext;

    // Show the dialog
    if (SUCCEEDED(hr)) {
        // Set the filter options
        if (nfilters > 0) {
            // Set the filters, and then show the dialog
            auto filters = ToFilterSpec(nfilters, filterName, filterPattern);
            hr = dlg->SetFileTypes(filters.size(), filters.data());
            if (SUCCEEDED(hr)) {
                ShowUserSaveDialog(dlg, name, bufferSize, numChars);
            }
        }
        else {
            // No filters, just show the dialog
            ShowUserSaveDialog(dlg, name, bufferSize, numChars);
        }
        dlg->Release();
    }
    // End Window-Specific Code
}



void show_browse_folder_dialog_c(char *buffer, int bufferSize, int *numChars)
{
    // Initialize outputs
    *numChars = 0;

    // Windows-Specific Code

    // Create the file dialog
    IFileDialog *dlg;
    HRESULT hr = CoCreateInstance(CLSID_FileOpenDialog, NULL, 
        CLSCTX_ALL, IID_IFileOpenDialog, reinterpret_cast<void**>(&dlg));
    if (SUCCEEDED(hr)) {
        // Establish options
        DWORD opt;
        hr = dlg->GetOptions(&opt);
        if (SUCCEEDED(hr)) {
            hr = dlg->SetOptions(opt | FOS_PICKFOLDERS);
            if (SUCCEEDED(hr)) {
                hr = dlg->Show(NULL);
                if (SUCCEEDED(hr)) {
                    IShellItem *pItem;
                    hr = dlg->GetResult(&pItem);
                    if (SUCCEEDED(hr)) {
                        PWSTR path;
                        hr = pItem->GetDisplayName(SIGDN_FILESYSPATH, &path);

                        size_t lstr = wcslen(path);
                        *numChars = (int)wcstombs(buffer, path, MIN(bufferSize, lstr));

                        CoTaskMemFree(path);
                    }
                    pItem->Release();
                }
            }
        }
        dlg->Release();
    }
    // End Window-Specific Code
}


int show_message_box_c(void *parent, const char *txt, const char *title, 
    int buttons, int icon)
{
    // Windows-Specific Code
    HWND hWnd = (HWND)parent;
    UINT btn, icn, type;
    int result;
    size_t nchars;
    wchar_t txtBuffer[4096], titleBuffer[4096];

    switch (buttons) {
        case MSGBX_BTN_OK:
            btn = MB_OK;
            break;

        case MSGBX_BTN_OK_CANCEL:
            btn = MB_OKCANCEL;
            break;

        case MSGBX_BTN_YES_NO:
            btn = MB_YESNO;
            break;

        case MSGBX_BTN_YES_NO_CANCEL:
            btn = MB_YESNOCANCEL;
            break;

        case MSGBX_BTN_RETRY_CANCEL:
            btn = MB_RETRYCANCEL;
            break;

        case MSGBX_BTN_ABORT_RETRY_IGNORE:
            btn = MB_ABORTRETRYIGNORE;
            break;

        case MSGBX_BTN_CANCEL_RETRY_CONTINUE:
            btn = MB_CANCELTRYCONTINUE;
            break;

        default:
            btn = MSGBX_NO_BUTTON;
            break;
    }

    switch (icon) {
        case MSGBX_ICON_ERROR:
            icn = MB_ICONERROR;
            break;

        case MSGBX_ICON_WARNING:
            icn = MB_ICONWARNING;
            break;

        case MSGBX_ICON_QUESTION:
            icn = MB_ICONQUESTION;
            break;

        case MSGBX_ICON_INFORMATION:
            icn = MB_ICONINFORMATION;
            break;

        default:
            icn = MSGBX_NO_ICON;
            break;
    }

    if (btn == MSGBX_NO_BUTTON && icn == MSGBX_NO_ICON) {
        type = 0;
    }
    else if (btn == MSGBX_NO_BUTTON && icn != MSGBX_NO_ICON) {
        type = icn;
    }
    else if (btn != MSGBX_NO_BUTTON && icn == MSGBX_NO_ICON) {
        type = btn;
    }
    else {
        type = btn | icn;
    }

    nchars = mbstowcs(txtBuffer, txt, MIN(4096, strlen(txt) + 1));
    nchars = mbstowcs(titleBuffer, title, MIN(4096, strlen(title) + 1));
    result = MessageBox(hWnd, txtBuffer, titleBuffer, type);

    switch (result) {
        case IDABORT:
            result = DIALOG_RESULT_ABORT;
            break;

        case IDCANCEL:
            result = DIALOG_RESULT_CANCEL;
            break;

        case IDCONTINUE:
            result = DIALOG_RESULT_CONTINUE;
            break;

        case IDIGNORE:
            result = DIALOG_RESULT_IGNORE;
            break;

        case IDNO:
            result = DIALOG_RESULT_NO;
            break;

        case IDOK:
            result = DIALOG_RESULT_OK;
            break;

        case IDRETRY:
            result = DIALOG_RESULT_RETRY;
            break;

        case IDYES:
            result = DIALOG_RESULT_YES;
            break;
    }

    // End Window-Specific Code

    // End
    return result;
}

File: src/ui_dialogs.f90
! ui_dialogs.f90

!> @brief This module provides means for displaying GUI type dialogs.
module ui_dialogs
    use iso_c_binding
    use iso_fortran_env
    use strings
    implicit none
    private
    public :: DIALOG_RESULT_CANCEL
    public :: DIALOG_RESULT_OK
    public :: DIALOG_RESULT_YES
    public :: DIALOG_RESULT_NO
    public :: DIALOG_RESULT_RETRY
    public :: DIALOG_RESULT_ABORT
    public :: DIALOG_RESULT_IGNORE
    public :: DIALOG_RESULT_CONTINUE
    public :: MSGBX_NO_ICON
    public :: MSGBX_NO_BUTTON
    public :: MSGBX_ICON_ERROR
    public :: MSGBX_ICON_WARNING
    public :: MSGBX_ICON_QUESTION
    public :: MSGBX_ICON_INFORMATION
    public :: MSGBX_BTN_OK
    public :: MSGBX_BTN_OK_CANCEL
    public :: MSGBX_BTN_YES_NO
    public :: MSGBX_BTN_YES_NO_CANCEL
    public :: MSGBX_BTN_RETRY_CANCEL
    public :: MSGBX_BTN_ABORT_RETRY_IGNORE
    public :: MSGBX_BTN_CANCEL_RETRY_CONTINUE
    public :: dialog_result
    public :: text_output_dialog_result
    public :: file_filter
    public :: initialize_ui_environment
    public :: clean_up_ui_environment
    public :: show_open_file_dialog
    public :: show_open_multifile_dialog
    public :: show_save_file_dialog
    public :: show_browse_folder_dialog
    public :: show_message_box

    ! TO DO: Message-Box type dialog & associated icons, buttons, etc.

! ------------------------------------------------------------------------------
    !> @brief The dialog result is a cancel button press.
    integer(int32), parameter :: DIALOG_RESULT_CANCEL = 0
    !> @brief The dialog result is an OK button press.
    integer(int32), parameter :: DIALOG_RESULT_OK = 1
    !> @brief The dialog result is a yes button press.
    integer(int32), parameter :: DIALOG_RESULT_YES = 2
    !> @brief The dialog result is a no button press.
    integer(int32), parameter :: DIALOG_RESULT_NO = 3
    !> @brief The dialog result is a retry button press.
    integer(int32), parameter :: DIALOG_RESULT_RETRY = 4
    !> @brief The dialog result is an abort button press.
    integer(int32), parameter :: DIALOG_RESULT_ABORT = 5
    !> @brief The dialog result is an ignore button press.
    integer(int32), parameter :: DIALOG_RESULT_IGNORE = 6
    !> @brief The dialog result is a continue button press.
    integer(int32), parameter :: DIALOG_RESULT_CONTINUE = 7
    !> @brief Indicates that no icon should be used for the dialog.
    integer(int32), parameter :: MSGBX_NO_ICON = 0
    !> @brief Indicates that no button should be used for the dialog.
    integer(int32), parameter :: MSGBX_NO_BUTTON = 0
    !> @brief Indicates that an error icon should be used for the dialog.
    integer(int32), parameter :: MSGBX_ICON_ERROR = 1000
    !> @brief Indicates that a warning icon should be used for the dialog.
    integer(int32), parameter :: MSGBX_ICON_WARNING = 1001
    !> @brief Indicates that a question icon should be used for the dialog.
    integer(int32), parameter :: MSGBX_ICON_QUESTION = 1002
    !> @brief Indicates that an information icon should be used for the dialog.
    integer(int32), parameter :: MSGBX_ICON_INFORMATION = 1003
    !> @brief Indicates that an OK button should be used.
    integer(int32), parameter :: MSGBX_BTN_OK = 1004
    !> @brief Indicates that an OK-Cancel button pair should be used.
    integer(int32), parameter :: MSGBX_BTN_OK_CANCEL = 1005
    !> @brief Indicates that a Yes-No button pair should be used.
    integer(int32), parameter :: MSGBX_BTN_YES_NO = 1006
    !> @brief Indicates that a Yes-No-Cancel button set should be used.
    integer(int32), parameter :: MSGBX_BTN_YES_NO_CANCEL = 1007
    !> @brief Indicates that a Retry-Cancel button set should be used.
    integer(int32), parameter :: MSGBX_BTN_RETRY_CANCEL = 1008
    !> @brief Indicates that an Abory-Retry-Ignore button set should be used.
    integer(int32), parameter :: MSGBX_BTN_ABORT_RETRY_IGNORE = 1009
    !> @brief Indicates that a Cancel-Retry-Continue button set should be used.
    integer(int32), parameter :: MSGBX_BTN_CANCEL_RETRY_CONTINUE = 1010

! ------------------------------------------------------------------------------
    !> @brief A type containing results from the dialog interaction.
    type dialog_result
        !> @brief The result flag.
        integer(int32) :: result
    end type

! ------------------------------------------------------------------------------
    !> @brief A type containing results from a dialog that returns text such
    !! as an open-file dialog.
    type, extends(dialog_result) :: text_output_dialog_result
        !> @brief A list of text returned by the dialog
        type(string), allocatable, dimension(:) :: string_list
    end type

! ------------------------------------------------------------------------------
    !> @brief Defines a file-type filter.
    type file_filter
        !> @brief The filter name.
        character(len = :), allocatable :: name
        !> @brief The filter pattern.
        character(len = :), allocatable :: pattern
    end type

! ------------------------------------------------------------------------------
    interface
        !> @brief Initializes the UI environment thereby allowing utilization
        !! of the dialogs.  This call must be made prior to calling and dialog
        !! routine.
        !!
        !! @return Returns true if the initialization was successful; else, 
        !! false.
        function init_ui_env_c() bind(C, name = "init_ui_env_c") result(rst)
            use iso_c_binding
            logical(c_bool) :: rst
        end function

        !> @brief Cleans up and finalizes the UI environment.  This should be
        !! called after all UI interactions are complete.
        subroutine clean_up_ui_environment() bind(C, name = "clean_up_ui_env_c")
            ! No arguments
        end subroutine

        !> @brief Shows the user an Open-File dialog.  Notice, the UI 
        !! environment must be initialized prior to calling this routine.
        !!
        !! @param[out] nameBuffer A buffer to which the user-selected 
        !!  filename will be written.
        !! @param[in] bufferSize The size of @p nameBuffer.
        !! @param[out] numChars The number of characters actually written
        !!  to @p nameBuffer.  Notice, if the user presses cancel, or does
        !!  not select any file this value will be 0.
        !! @param[in] nfilters The number of filter strings.
        !! @param[in] filterName An array containing the filter names.
        !! @param[in] filterPattern An array containing the filter pattern.
        subroutine show_open_file_dialog_c(nameBuffer, bufferSize, numChars, &
                nfilters, filterName, filterPattern) &
                bind(C, name = "show_open_file_dialog_c")
            use iso_c_binding
            character(kind = c_char), intent(out) :: nameBuffer(*)
            integer(c_int), intent(in), value :: bufferSize, nfilters
            integer(c_int), intent(out) :: numChars
            type(c_ptr), intent(in) :: filterName(nfilters), &
                filterPattern(nfilters)
        end subroutine

        !> @brief Shows the user an Open-File dialog capable of returning 
        !! multiple files.  Notice, the UI environment must be initialized prior
        !! to calling this routine.
        !!
        !! @param[out] nameBuffers An array of string buffers where the !!
        !!  filenames will be written.
        !! @param[in] numBuffers The number of buffers available.
        !! @param[in] bufferSize The size of each buffer.
        !! @param[out] numNames The number of names actually written.
        !! @param[out] numChars An array containing the number of characters 
        !!  written in each buffer.
        !! @param[in] nfilters The number of filter strings.
        !! @param[in] filterName An array containing the filter names.
        !! @param[in] filterPattern An array containing the filter pattern.
        subroutine show_multi_file_open_file_dialog_c(nameBuffers, numBuffers, &
                bufferSize, numNames, numChars, nfilters, filterName, &
                filterPattern) &
                bind(C, name = "show_multi_file_open_file_dialog_c")
            use iso_c_binding
            integer(c_int), intent(in), value :: numBuffers, bufferSize, &
                nfilters
            type(c_ptr), intent(out) :: nameBuffers(numBuffers)
            integer(c_int), intent(out) :: numNames, numChars(numBuffers)
            type(c_ptr), intent(in) :: filterName(nfilters), &
                filterPattern(nfilters)
        end subroutine

        !> @brief Shows the user a Save-File dialog.  Notice, the UI environment
        !! must be initialized prior to calling this routine.
        !!
        !! @param[in] defaultExt The default file extension to append.
        !! @param[in] nfilters The number of filter strings.
        !! @param[in] filterName An array containing the filter names.
        !! @param[in] filterPattern An array containing the filter pattern.
        !! @param[out] name The user-defined filename.
        !! @param[in] bufferSize the available size of @p name.
        !! @param[out] numChars The actual number of characters written to
        !!  @p name.
        subroutine show_save_file_dialog_c(defaultExt, nfilters, filterName, &
                filterPattern, name, bufferSize, numChars) &
                bind(C, name = "show_save_file_dialog_c")
            use iso_c_binding
            character(kind = c_char), intent(in) :: defaultExt(*)
            integer(c_int), intent(in), value :: nfilters, bufferSize
            type(c_ptr), intent(in) :: filterName(nfilters), &
                filterPattern(nfilters)
            character(kind = c_char), intent(out) :: name(*)
            integer(c_int), intent(out) :: numChars
        end subroutine

        !> @brief Shows the user a Browse-Folder dialog.  Notice, the UI 
        !! environment must be initialized prior to calling this routine.
        !!
        !! @param[out] buffer A buffer to which the output will be written.
        !! @param[in] bufferSize The available size of @p buffer.
        !! @param[out] numChars The actual number of characters written to
        !!  @p buffer.
        subroutine show_browse_folder_dialog_c(buffer, bufferSize, numChars) &
                bind(C, name = "show_browse_folder_dialog_c")
            use iso_c_binding
            character(kind = c_char), intent(out) :: buffer(*)
            integer(c_int), intent(in), value :: bufferSize
            integer(c_int), intent(out) :: numChars
        end subroutine

        !> @brief Shows the user a message box.
        !!
        !! @param[in] parent A pointer to the parent window.  This can be null
        !!  in the event there isn't a parent window.
        !! @param[in] txt The message box text.
        !! @param[in] title The title of the message box.
        !! @param[in] buttons
        !! @param[in] icon
        !! 
        !! @return The dialog result.
        function show_message_box_c(parent, txt, title, buttons, icon) &
                result(rst) bind(C, name = "show_message_box_c")
            use iso_c_binding
            type(c_ptr), intent(in), value :: parent
            character(kind = c_char), intent(in) :: txt(*), title(*)
            integer(c_int), intent(in), value :: buttons, icon
            integer(c_int) :: rst
        end function
    end interface

! ------------------------------------------------------------------------------

contains
! ------------------------------------------------------------------------------
    !> @brief Initializes the UI environment thereby allowing utilization
    !! of the dialogs.  This call must be made prior to calling and dialog
    !! routine.
    !!
    !! @return Returns true if the initialization was successful; else, 
    !! false.
    function initialize_ui_environment() result(rst)
        ! Arguments
        logical :: rst

        ! Process
        rst = logical(init_ui_env_c())
    end function

! ------------------------------------------------------------------------------
    !> @brief Shows the user an Open-File dialog capable of single-file 
    !! selection.
    !!
    !! @param[in] filters An optional array containing file filter information.
    !!
    !! @return The results from the dialog.
    function show_open_file_dialog(filters) result(rst)
        ! Arguments
        class(file_filter), intent(in), dimension(:), optional :: filters
        type(text_output_dialog_result) :: rst

        ! Parameters
        integer(c_int), parameter :: bufferSize = 2048
        integer(c_int), parameter :: maxFilters = 100

        ! Local Variables
        integer(c_int) :: i, nChars, nfilters
        character(kind = c_char, len = bufferSize) :: buffer
        type(c_ptr) :: nullarray(1), namePtrs(maxFilters), &
            patternPtrs(maxFilters)
        character(len = 1024), dimension(maxFilters), target :: nameList, &
            patternList

        ! Initialization
        nullarray = c_null_ptr

        ! Show the user the dialog
        if (present(filters)) then
            nfilters = min(size(filters), maxFilters)
            do i = 1, nfilters
                nameList(i) = filters(i)%name // C_NULL_CHAR
                patternList(i) = filters(i)%pattern // C_NULL_CHAR

                namePtrs(i) = c_loc(nameList(i))
                patternPtrs(i) = c_loc(patternList(i))
            end do
            call show_open_file_dialog_c(buffer, bufferSize, nChars, nfilters, &
                namePtrs(1:nfilters), patternPtrs(1:nfilters))
        else
            call show_open_file_dialog_c(buffer, bufferSize, nChars, 0, &
                nullarray, nullarray)
        end if

        ! Copy the buffer back to a Fortran-friendly string
        if (nChars > 0) then
            rst%result = DIALOG_RESULT_OK
            allocate(rst%string_list(1))
            rst%string_list(1)%str = to_fortran_string(buffer, nChars)
        else
            rst%result = DIALOG_RESULT_CANCEL
            allocate(rst%string_list(0))
        end if
    end function

! ------------------------------------------------------------------------------
    !> @brief Shows the user an Open-File dialog capable of multiple-file 
    !! selection.
    !!
    !! @param[in] filters An optional array containing file filter information.
    !!
    !! @return The results from the dialog.
    function show_open_multifile_dialog(filters) result(rst)
        ! Arguments
        ! Arguments
        class(file_filter), intent(in), dimension(:), optional :: filters
        type(text_output_dialog_result) :: rst

        ! Parameters
        integer(c_int), parameter :: numBuffers = 1000
        integer(c_int), parameter :: bufferSize = 2048
        integer(c_int), parameter :: maxFilters = 100

        ! Local Variables
        integer(c_int) :: i, nFilters, nNames, nChars(numBuffers)
        character(len = bufferSize), dimension(numBuffers), target :: buffer
        type(c_ptr) :: nullarray(1), namePtrs(maxFilters), &
            patternPtrs(maxFilters), bufferPtrs(numBuffers)
        character(len = 1024), dimension(maxFilters), target :: nameList, &
            patternList

        ! Initialization
        nullarray = C_NULL_PTR
        do i = 1, numBuffers
            bufferPtrs(i) = c_loc(buffer(i))
        end do

        ! Show the user the dialog
        if (present(filters)) then
            nFilters = min(size(filters), maxFilters)
            do i = 1, nFilters
                nameList(i) = filters(i)%name // C_NULL_CHAR
                patternList(i) = filters(i)%pattern // C_NULL_CHAR

                namePtrs(i) = c_loc(nameList(i))
                patternPtrs(i) = c_loc(patternList(i))
            end do
            call show_multi_file_open_file_dialog_c(bufferPtrs, numBuffers, &
                bufferSize, nNames, nChars, nFilters, namePtrs(1:nFilters), &
                patternPtrs(1:nFilters))
        else
            call show_multi_file_open_file_dialog_c(bufferPtrs, numBuffers, &
                bufferSize, nNames, nChars, 0, nullarray, nullarray)
        end if

        ! Collect the output from the buffer
        if (nNames > 0) then
            rst%result = DIALOG_RESULT_OK
            allocate(rst%string_list(nNames))
            do i = 1, nNames
                rst%string_list(i)%str = to_fortran_string(buffer(i), nChars(i))
            end do
        else
            rst%result = DIALOG_RESULT_CANCEL
            allocate(rst%string_list(0))
        end if
    end function

! ------------------------------------------------------------------------------
    !> @brief Shows the user a Save-File dialog.
    !!
    !! @param[in] defaultExt The default extension to append to the filename in 
    !!  the event the user does not specify an extension (e.g. "txt").
    !! @param[in] filters An array containing file filter information.
    !!
    !! @return The results from the dialog.
    function show_save_file_dialog(defaultExt, filters) result(rst)
        ! Arguments
        character(len = *), intent(in) :: defaultExt
        class(file_filter), intent(in), dimension(:), optional :: filters
        type(text_output_dialog_result) :: rst

        ! Parameters
        integer(c_int), parameter :: bufferSize = 2048
        integer(c_int), parameter :: maxFilters = 100

        ! Local Variables
        integer(c_int) :: i, nchars, nfilters
        character(kind = c_char, len = bufferSize) :: buffer
        type(c_ptr) :: nullarray(1), namePtrs(maxFilters), &
            patternPtrs(maxFilters)
        character(len = 1024), dimension(maxFilters), target :: nameList, &
            patternList
        character(kind = c_char, len = :), allocatable :: ext

        ! Initialization
        nullarray = C_NULL_PTR
        ext = defaultExt // C_NULL_CHAR

        ! Show the user the dialog
        if (present(filters)) then
            nfilters = min(size(filters), maxFilters)
            do i = 1, nfilters
                nameList(i) = filters(i)%name // C_NULL_CHAR
                patternList(i) = filters(i)%pattern // C_NULL_CHAR

                namePtrs(i) = c_loc(nameList(i))
                patternPtrs(i) = c_loc(patternList(i))
            end do
            call show_save_file_dialog_c(ext, nfilters, namePtrs(1:nfilters), &
                patternPtrs(1:nfilters), buffer, bufferSize, nchars)
        else
            call show_save_file_dialog_c(ext, 0, nullarray, nullarray, buffer, &
                bufferSize, nchars)
        end if

        ! Collect the output from the buffer
        if (nchars > 0) then
            rst%result = DIALOG_RESULT_OK
            allocate(rst%string_list(1))
            rst%string_list(1)%str = to_fortran_string(buffer, nchars)
        else
            rst%result = DIALOG_RESULT_CANCEL
            allocate(rst%string_list(0))
        end if
    end function

! ------------------------------------------------------------------------------
    !> @brief Shows the user a dialog allowing them to browse to a folder.
    !!
    !! @return The results from the dialog.
    function show_browse_folder_dialog() result(rst)
        ! Arguments
        type(text_output_dialog_result) :: rst

        ! Parameters
        integer(c_int), parameter :: bufferSize = 2048

        ! Local Variables
        integer(c_int) :: nchars
        character(kind = c_char, len = bufferSize) :: buffer

        ! Process
        call show_browse_folder_dialog_c(buffer, bufferSize, nchars)
        if (nchars > 0) then
            rst%result = DIALOG_RESULT_OK
            allocate(rst%string_list(1))
            rst%string_list(1)%str = to_fortran_string(buffer, nchars)
        else
            rst%result = DIALOG_RESULT_CANCEL
            allocate(Rst%string_list(0))
        end if
    end function

! ------------------------------------------------------------------------------
    !> @brief Shows the user a message box dialog.
    !!
    !! @param[in] txt The text to display in the message box.
    !! @param[in] title The title to display in the header of the message box.
    !! @param[in] buttons Defines what button group to display.  The following
    !!  choices are acceptable: MSGBX_BTN_OK, MSGBX_BTN_OK_CANCEL, 
    !!  MSGBX_BTN_YES_NO, MSGBX_BTN_YES_NO_CANCEL, MSGBX_BTN_RETRY_CANCEL, 
    !!  MSGBX_BTN_ABORT_RETRY_IGNORE, MSGBX_BTN_CANCEL_RETRY_CONTINUE.
    !! @param[in] icon Defines what icons to display.  The following choices
    !!  are acceptable: MSGBX_NO_ICON, MSGBX_NO_BUTTON, MSGBX_ICON_ERROR, 
    !!  MSGBX_ICON_WARNING, MSGBX_ICON_QUESTION, MSGBX_ICON_INFORMATION.
    !! @param[in] parent An optional argument defining a pointer to the parent
    !!  window.  If not supplied, no parent is assumed.
    !!
    !! @return The results from the dialog.
    function show_message_box(txt, title, buttons, icon, parent) result(rst)
        ! Arguments
        character(len = *), intent(in) :: txt, title
        integer(int32), intent(in) :: buttons, icon
        type(c_ptr), optional :: parent
        type(dialog_result) :: rst

        ! Local Variables
        character(kind = c_char, len = :), allocatable :: ctxt, ctitle
        type(c_ptr) :: parentPtr

        ! Initialization
        ctxt = txt // C_NULL_CHAR
        ctitle = title // C_NULL_CHAR
        if (present(parent)) then
            parentPtr = parent
        else
            parentPtr = c_null_ptr
        end if

        ! Process
        rst%result = show_message_box_c(parentPtr, ctxt, ctitle, buttons, icon)
    end function

! ------------------------------------------------------------------------------
end module
